### ISSUE TITLE
Feature request: first-class AMD ROCm 7.14 support on Windows

### ISSUE THREAD

**Issue by qweqweewqe7-create** [sorryhyun/anima_lora#88]
## Summary

I have already been using a locally modified version of `anima_lora` on an AMD Radeon GPU with ROCm, and the core training path works much better than I initially expected.

This is therefore not only a theoretical compatibility request: **the core LoRA training path, backward, optimizer step, and `torch.compile` have all worked locally on ROCm**. The remaining work appears to be mostly installation/dependency/backend selection rather than a large model-code port.

I think it would be practical to add a first-class AMD path that automatically installs/selects **ROCm 7.14** while preserving the current CUDA path for NVIDIA users.

## Tested environment

- OS: Windows 11
- GPU: AMD Radeon RX 9070 XT (`gfx1201`)
- Backend: ROCm 7.14 PyTorch environment
- Anima LoRA: locally modified from this repository

I have also been using AMD ROCm PyTorch with the same general Anima/ComfyUI stack, so the basic model operators are already known to work on this GPU family.

### 1. Core training

The normal training path works locally on ROCm, including:

- model loading
- forward pass
- loss computation
- backward pass
- AdamW optimizer step
- LoRA training

The project already relies mostly on normal PyTorch device APIs (`torch.cuda.*`, `device="cuda"`, etc.). ROCm PyTorch intentionally exposes the CUDA-compatible PyTorch API, so most of this code does not need an AMD-specific rewrite.

### 2. PyTorch SDPA works well

For AMD, `attn_mode = "torch"` is the best default in my testing.

The existing `networks/attention_dispatch.py` already has a clean PyTorch SDPA path using:

```python
torch.nn.functional.scaled_dot_product_attention(...)
```

so no new attention implementation is required just to support ROCm.

### 3. `torch.compile` works and should stay enabled

This is important: **I tested `torch.compile` on ROCm and it works in training.**

It also provides a meaningful memory reduction in my local tests, so I do **not** recommend disabling it for the AMD preset/backend.

Suggested AMD default:

```toml
attn_mode = "torch"
torch_compile = true
```

The project currently treats `torch.compile` as a major part of its VRAM/performance strategy, and ROCm does not require falling back to eager mode here in my tested setup.

## Flash Attention on AMD

I also tried an AMD-compatible Flash Attention implementation.

It worked, but for this Anima training workload it was approximately **3x slower than PyTorch SDPA** on my system.

Because of that, I would not make Flash Attention an AMD dependency or default. It only adds installation complexity while performing substantially worse in this case.

My recommendation is:

```text
NVIDIA -> flash (current default)
AMD    -> torch SDPA
```

AMD users could still manually experiment with other attention backends later, but the official ROCm path should stay simple and use SDPA by default.

## Proposed backend selection

Something like this would keep the current behavior while allowing an explicit override:

```text
ANIMA_BACKEND=auto   # default
ANIMA_BACKEND=cuda
ANIMA_BACKEND=rocm
```

Possible `auto` behavior:

```text
NVIDIA GPU -> CUDA path
AMD GPU    -> ROCm path
```

For unusual mixed-vendor systems, an explicit environment variable can override detection.

## Windows installer changes

The current `install.ps1` is strongly CUDA-specific: it checks for CUDA 13.2 and can download/install the CUDA toolkit automatically.

For a ROCm backend, I suggest branching before that step:

```powershell
if ($Backend -eq "cuda") {
    # existing CUDA 13.2 path
}
elseif ($Backend -eq "rocm") {
    # skip CUDA toolkit installation
    # install/use ROCm 7.14 PyTorch packages
}
```

The AMD path should automatically use the official ROCm 7.14 Windows PyTorch package source rather than requiring users to manually replace Torch after installation.

## `pyproject.toml` / `uv` is probably the main packaging issue

This looks like the most important part to solve correctly.

At the moment the project contains NVIDIA-specific dependencies/settings in the normal environment, including:

- CUDA 13.2 PyTorch index
- CUDA-specific prebuilt `flash-attn` wheels
- Windows Triton/CUDA-related dependencies

Simply installing ROCm Torch first is not enough if a later `uv sync` resolves the project back to the CUDA Torch packages.

So I think the clean implementation is to separate common dependencies from backend-specific dependencies/resolution:

```text
common
  - accelerate
  - transformers
  - diffusers
  - safetensors
  - etc.

cuda
  - CUDA PyTorch build
  - flash-attn CUDA wheel
  - CUDA/Triton-specific dependencies

rocm
  - ROCm 7.14 PyTorch build
  - no flash-attn by default
```

The important requirement is that later `uv run` / `uv sync` operations must not silently replace the ROCm PyTorch installation with the CUDA build.

## Suggested AMD defaults

I would start with only these backend-specific behavioral differences:

```toml
# AMD / ROCm
attn_mode = "torch"
torch_compile = true
gradient_checkpointing = false  # unchanged unless the selected preset needs it
optimizer_type = "AdamW"
```

Most of the rest of the existing configuration can remain shared with NVIDIA.

## GUI GPU monitor

`gui/gpu_status.py` currently uses `nvidia-smi`.

This does not appear to be a blocker because the widget already hides itself if `nvidia-smi` is unavailable. AMD monitoring can therefore be added later without blocking initial ROCm support.

## Scope I would recommend for an initial ROCm implementation

1. Add backend selection (`auto` / `cuda` / `rocm`).
2. Detect AMD vs NVIDIA for the default backend.
3. Keep the current CUDA 13.2 installation path unchanged for NVIDIA.
4. Add a ROCm 7.14 installation path for AMD.
5. Prevent CUDA-specific Torch / Flash Attention packages from being installed in the ROCm environment.
6. Default to `attn_mode = "torch"` on AMD.
7. Keep `torch_compile = true` on AMD.
8. Add a small post-install smoke test, for example:

```python
import torch

print("torch:", torch.__version__)
print("HIP:", torch.version.hip)
print("GPU:", torch.cuda.get_device_name(0))

assert torch.cuda.is_available()
assert torch.version.hip is not None
```

9. Treat AMD GPU monitoring / optional alternative attention backends as follow-up work rather than blockers.

## Why I think this is worth supporting

The biggest compatibility concerns I expected were attention backward and `torch.compile`, but in actual local use the important training path already works on ROCm.

The remaining gap is mostly making installation reproducible and preventing the CUDA-specific dependency setup from taking over the environment.

So this looks like a relatively small first-class backend addition compared with a full port.

I can also provide additional logs/benchmarks from the working ROCm setup if useful. I have already compared AMD Flash Attention against PyTorch SDPA and confirmed that SDPA is the preferable default on this workload.

### Current local conclusions

| Item | ROCm result |
|---|---|
| Core LoRA training | Works |
| Forward | Works |
| Backward | Works |
| AdamW step | Works |
| PyTorch SDPA | Works / preferred |
| AMD Flash Attention | Works, but ~3x slower than SDPA in my test |
| `torch.compile` | Works |
| `torch.compile` memory benefit | Confirmed |
| Recommended AMD attention default | `torch` |
| Recommended AMD compile default | enabled |

I have not exhaustively validated every optional training/inference feature in the repository, so I would initially describe this as support for the **core Anima LoRA training path on ROCm 7.14**, with optional methods/features validated incrementally.

**Comment by sorryhyun** [comment]
Thanks. I think it's almost there but I cannot test rocm path so can you make PR for this if you can do any? I think you already listed what to test so I expect most concerns would be resolved. By the way how did you run FA2 in rocm path? In official repo they recommended 2 ways as [here](https://github.com/dao-ailab/flash-attention#triton-backend)
